riscv-runtime/src/hash.rs

```diff
@@ -47,40 +47,72 @@ pub fn keccakf(input: &[u64; 25], output: &mut [u64; 25]) {
     }
 }
 
-// Output number of bytes for keccak-256 (32 bytes)
-const W: usize = 32;
-
-/// Keccak function that calls the keccakf machine.
-/// Input is a byte array of arbitrary length and a delimiter byte.
-/// Output is a byte array of length W.
-pub fn keccak(data: &[u8], delim: u8) -> [u8; W] {
-    let mut b = [[0u8; 200]; 2];
-    let [mut b_input, mut b_output] = &mut b;
-    let rate = 200 - (2 * W);
-    let mut pt = 0;
-
-    // update
-    for &byte in data {
-        b_input[pt] ^= byte;
-        pt = (pt + 1) % rate;
-        if pt == 0 {
-            unsafe {
-                keccakf(mem::transmute(&b_input), mem::transmute(&mut b_output));
+pub struct Keccak {
+    state: [u64; 25],
+    buffer: [u8; 136],
+    buffer_pos: usize,
+    rate: usize,
+}
+
+impl Keccak {
+    pub fn v256() -> Self {
+        //RATE = 136
+
+        Self {
+            state: [0u64; 25],
+            buffer: [0u8; 136],
+            buffer_pos: 0,
+            rate: 136,
+        }
+    }
+
+    pub fn update(&mut self, data: &[u8]) {
+        let mut input = data;
+        while !input.is_empty() {
+            let space = self.rate - self.buffer_pos;
+            let to_absorb = space.min(input.len());
+
+            self.buffer[self.buffer_pos..self.buffer_pos + to_absorb]
+                .copy_from_slice(&input[..to_absorb]);
+            self.buffer_pos += to_absorb;
+            input = &input[to_absorb..];
+
+            if self.buffer_pos == self.rate {
+                self.absorb_block();
+                self.buffer_pos = 0;
             }
-            mem::swap(&mut b_input, &mut b_output);
         }
     }
 
-    // finalize
-    b_input[pt] ^= delim;
-    b_input[rate - 1] ^= 0x80;
-    unsafe {
-        keccakf(mem::transmute(&b_input), mem::transmute(&mut b_output));
+    fn absorb_block(&mut self) {
+        let rate_words = self.rate / 8;
+        for i in 0..rate_words {
+            let chunk = &self.buffer[i * 8..(i + 1) * 8];
+            let val = u64::from_le_bytes(chunk.try_into().unwrap());
+            // Note: lo/hi positions are swapped to match the Keccak machine memory layout
+            let swapped = ((val << 32) & 0xFFFFFFFF00000000) | ((val >> 32) & 0x00000000FFFFFFFF);
```

Comment:
Instead of swapping, it would probably be better to define the state as `[u32; _]` instead of `[u64; _]`, but I guess that would require changing the `keccakf` interface... which I don't know why it would use `[u64; _]` in the first place, instead of `[u32; _]` (which is the most natural choice, given the 4-byte alignment required of the machine). Maybe it was copied from Poseidon? But `poseidon_gl` has a reason to be `u64`, as it operates on field elements.